markdowns/20_REST_API.md

```diff
@@ -0,0 +1,370 @@
+
+```yaml
+pipelines:
+  - name: query 
+    nodes:
+      - name: Retriever
+        inputs: [Query]
+```
+
+2. Create an Indexing Pipeline
+
+You can define an indexing pipeline in the same pipeline YAML file and index your documents to Elasticsearch through REST API. For that, create `FileTypeClassifier`, `TextConverter`, `PDFToTextConverter`, and last, `PreProcessor` nodes with parameters to split documents. 
+
+```yaml
+components:
+    ...
+  - name: FileTypeClassifier
+    type: FileTypeClassifier
+  - name: TextFileConverter
+    type: TextConverter
+  - name: PDFFileConverter
+    type: PDFToTextConverter
+  - name: Preprocessor
+    type: PreProcessor
+    params:
+      split_by: word
+      split_length: 1000
+      split_overlap: 50 
+      split_respect_sentence_boundary: True 
+```
+
+Then, go below to the `pipelines` section of the YAML file and create a new pipeline called `indexing`. In this pipeline, indicate how these nodes are connected to each other, Retriever, and DocumentStore. This indexing pipeline supports `.txt` and `.pdf` files and pre-processes them before loading to the Elasticsearch.
+
+```yaml
+pipelines:
+  ...
+  - name: indexing
+    nodes:
+      - name: FileTypeClassifier
+        inputs: [File]
+      - name: TextFileConverter
+        inputs: [FileTypeClassifier.output_1]
+      - name: PDFFileConverter
+        inputs: [FileTypeClassifier.output_2]
+      - name: Preprocessor
+        inputs: [PDFFileConverter, TextFileConverter]
+      - name: Retriever
+        inputs: [Preprocessor]
+      - name: DocumentStore
+        inputs: [Retriever]
+```
+
+After completing query and indexing pipelines, add `version: ignore` to the top of the file and the pipeline YAML file is ready to run by Haystack API.
+
+```yaml
+version: ignore
+
+components:
+  - name: DocumentStore
+    type: ElasticsearchDocumentStore
+  - name: Retriever
+    type: EmbeddingRetriever
+    params:
+      document_store: DocumentStore
+      top_k: 5 
+      embedding_model: sentence-transformers/multi-qa-mpnet-base-dot-v1
+  - name: FileTypeClassifier
+    type: FileTypeClassifier
+  - name: TextFileConverter
+    type: TextConverter
+  - name: PDFFileConverter
+    type: PDFToTextConverter
+  - name: Preprocessor
+    type: PreProcessor
+    params:
+      split_by: word
+      split_length: 1000
+      split_overlap: 50 
+      split_respect_sentence_boundary: True
+
+pipelines:
+  - name: query 
+    nodes:
+      - name: Retriever
+        inputs: [Query]
+  - name: indexing
+    nodes:
+      - name: FileTypeClassifier
+        inputs: [File]
+      - name: TextFileConverter
+        inputs: [FileTypeClassifier.output_1]
+      - name: PDFFileConverter
+        inputs: [FileTypeClassifier.output_2]
+      - name: Preprocessor
+        inputs: [PDFFileConverter, TextFileConverter]
+      - name: Retriever
+        inputs: [Preprocessor]
+      - name: DocumentStore
+        inputs: [Retriever]
+```
+
+## Start Haystack API
```

Comment:
Tbh, I explained how to run it with Docker in the first version as an alternative to gunicorn so that people know it's possible. However, I couldn't get a stable development environment with Docker (it couldn't connect to elasticsearch, gave errors etc.), which prevented me to interact easily with the pipeline YAML file. Since this tutorial is about how to interact with REST API, I wanted to focus on that part instead of explaining how to run it with the current `docker-compose.yml` file as there are some additional steps there.